### PR TITLE
Controls - CheckBox

### LINKED ISSUE #4: [Proposal] Checkbox
_From @jamesmontemagno on December 23, 2016 7:7_

## Proposal:
Implementation to have a checkbox with 2 states:
Selected, Unselected

## Properties
boolean: IsChecked: Gets or sets the checked state
boolean: IsEnabled: Gets or sets enabled status

## Events
OnCheckChanged: Triggered when check status changes.

## Implementation
Android: Built in control
iOS: Custom UIButton with graphics
UWP: Built in control

_Copied from original issue: FormsCommunityToolkit/Controls#1_

### PR DESCRIPTION
**Pull request by wcoder**
Fixes #4 .

Changes Proposed in this pull request:
- Added implementation `CheckBox` control (iOS, Android, UWP)
- Added documentation about `CheckBox`
- Added `Controls` projects to `Samples` solution
- Fixed `Samples` solution (References between `Animations` projects)
- Update Xamarin.Forms for projects of `Controls`

Based on:
- [XForms.Controls](https://github.com/XForms/XForms.Controls)
- [Forms.Controls](https://github.com/messier16/Forms.Controls)

### CODE REVIEW

File: src/Controls/Controls.iOS/Views/CheckBoxView.cs [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+namespace FormsCommunityToolkit.Controls.iOS.Views
+{
+    /// <summary>
+    /// Copyright (c) 2017 fferegrino (Antonio Feregrino)
+    /// https://github.com/messier16/Forms.Controls/blob/master/Forms.Controls.iOS/Native/Checkbox/Checkbox.cs
+    /// </summary>
+    [Register("CheckBoxView")]
+    public class CheckBoxView : UIControl
+    {
+        #region Color layers
```
Comment (bartlannoeye) [inline src/Controls/Controls.iOS/Views/CheckBoxView.cs]: I really dislike regions :)

File: src/Controls/Controls.iOS/Views/CheckBoxView.cs [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+                _checked = value;
+                ColorLayers();
+            }
+        }
+
+        #endregion
+
+        #region Internal and private propertoies
+
+        private readonly CAShapeLayer containerLayer = new CAShapeLayer();
```
Comment (bartlannoeye) [inline src/Controls/Controls.iOS/Views/CheckBoxView.cs]: privates have _

File: src/Controls/Controls/CheckBox.cs [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+	public class CheckBox : View
+	{
+		public static readonly BindableProperty IsCheckedProperty =
+			BindableProperty.Create(
+				nameof(IsChecked), typeof(bool), typeof(bool), false,
+				BindingMode.TwoWay);
+
+		public event EventHandler<bool> CheckedChanged;
+
+
```
Comment (bartlannoeye) [inline src/Controls/Controls/CheckBox.cs]: one newline should be plenty

**Comment by Depechie** [comment]
@wcoder 
The current implementation is working, so that is great :)
Could you maybe change the samples page a bit so that the control itself is centered on the demo page? Other than that it looks fine, only problem I have, we have no Controls admins on the current project. So not sure if people like the way this current version is created?
Personally last time I added a checkbox to Xamarin Forms is was not a box but a round version :/ so taste is very user dependent...

**Comment by owlstack** [comment]
Hi, I tested the UWP implementation and I found a minor bug that causes a crash (Exception). I will continue to test but once I'm all set, how can I submit a pull request to add the fix to prevent the exception? Thanks! (First time contributor, so nervous about breaking anything) :)

**Comment by Depechie** [comment]
@owlstack let me first see what the status is of this PR and when we merged in into the Dev branch, you'll be able to do a new PR with your code solving the actual issue.
I'll keep you posted, ok?

**Comment by wcoder** [comment]
Hi everyone

@Depechie 

I synchronized this branch with `dev` and was centered CheckBox on the demo page.

@owlstack

I have checked the solution and wasn't find the errors and crashes for projects.

**Comment by owlstack** [comment]
@wcoder There is a crash for UWP portion. Only a null check is required. Can I add the fix? Thanks!

**Comment by wcoder** [comment]
@owlstack surely you can add the fix, push your changes to branch: [wcoder:controls/checkbox](https://github.com/wcoder/FormsCommunityToolkit/tree/controls/checkbox)

https://github.com/wcoder/FormsCommunityToolkit/tree/controls/checkbox

File: src/Controls.sln [hunk 4]
```diff
@@ -1,9 +1,9 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio 14
-VisualStudioVersion = 14.0.25420.1
+# Visual Studio 15
```
Comment (skendrot) [inline src/Controls.sln]: Is there a concern with increasing the version? Are lower versions still needed?

File: src/Controls.sln [hunk 5]
```diff
@@ -1,9 +1,9 @@
...
+VisualStudioVersion = 15.0.27130.2026
 MinimumVisualStudioVersion = 10.0.40219.1
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Controls", "Controls\Controls\Controls.csproj", "{572B958E-CC33-47C1-95FA-1B3A0506F2ED}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Controls", "Controls\Controls\Controls.csproj", "{572B958E-CC33-47C1-95FA-1B3A0506F2ED}"
```
Comment (skendrot) [inline src/Controls.sln]: Why are project guids changing?

File: src/Controls/Controls.Android/CheckBoxRenderer.cs [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+using Android.Widget;
+using XamarinCommunityToolkit.Controls.Droid;
+using Xamarin.Forms;
+using Xamarin.Forms.Platform.Android;
+using CheckBox = XamarinCommunityToolkit.Controls.CheckBox;
+using NativeCheckBox = Android.Widget.CheckBox;
+
+[assembly: ExportRenderer(typeof(CheckBox), typeof(CheckBoxRenderer))]
+
+namespace XamarinCommunityToolkit.Controls.Droid
```
Comment (skendrot) [inline src/Controls/Controls.Android/CheckBoxRenderer.cs]: I know this is bigger than this one control, but I would recommend changing namespaces to Xamarin.Toolkit.Controls.Droid

File: src/Controls/Controls.Android/Controls.Android.csproj [hunk 7]
```diff
@@ -6,11 +6,11 @@
     <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
     <ProductVersion>8.0.30703</ProductVersion>
     <SchemaVersion>2.0</SchemaVersion>
-    <ProjectGuid>{88DC2010-BE3A-4B86-A72F-3DF68413858B}</ProjectGuid>
+    <ProjectGuid>{A62DCD64-5BD4-4E9A-ADF7-734C57D2F8A0}</ProjectGuid>
```
Comment (skendrot) [inline src/Controls/Controls.Android/Controls.Android.csproj]: Why are project guids changing?

**Comment by pauldipietro** [comment]
@wcoder I haven't gone through this yet, but this may be worth resubmitting to Forms core under https://github.com/xamarin/Xamarin.Forms/issues/1728. Have you seen that proposal?

**Comment by wcoder** [comment]
@pauldipietro I'm ready for resubmitting this control to Forms repository.

P.S. About this repository:
After changes named "Completely reorganize and update repo!" my enthusiasm became low for sync this PR.
Please, provide infrastructure for UI controls and Controls admins on the current project.

**Comment by WilliamABradley** [comment]
It would be really awesome if this integrated an Indeterminate Checkbox on Android, such as: https://github.com/sevar83/indeterminate-checkbox, that way the experience can be combined with UWP's Implementation :)

**Comment by pauldipietro** [comment]
Closing this for now; I think the focus should remain on XF Core as an F100 issue for a CheckBox is open (as referred to earlier) and something should be contributed there.

**Pull request closed without merging**